Upgrade to Django 2.2

millerdev opened the pull request:
Test failures are expected. Review by commit :blowfish:

millerdev committed:
Fix CaseTransaction not hashable

TypeError: unhashable type: 'CaseTransaction'

millerdev committed:
Fix location fixture tests

TypeError: Direct assignment to the forward side of a many-to-many set is prohibited. Use include_only.set() instead.

millerdev committed:
Upgrade django to 2.2 (run make requirements)

millerdev committed:
Eliminate test interaction

TestDomainUserHistory.test_calculate_users_in_all_domains failed when running tests locally due to other tests that did not clean up after themselves.

This is unrelated to the Django 2 upgrade.

esoergel commented on corehq/form_processor/models.py:
```diff
@@ -1395,6 +1395,9 @@ def __eq__(self, other):
         )
 
+    def __hash__(self):
+        return hash((self.case_id, self.type, self.form_id))
```
huh, can there not be multiple updates to a case in a single form?  This `__hash__` definition is consistent with `__eq__`, so it seems fine, but I didn't realize that.

esoergel reviewed (commented):
There's a build failure I haven't looked at, but the changes up to here look good.

millerdev committed:
Patch TestCase.databases to allow access

millerdev committed:
Django templates now require text (not bytes)

As of Django 2.1 it is no longer possible to use bytes as the source of a template object. Changed in https://github.com/django/django/commit/4dadf4eb7bad7c8e9154ff86e7e42ed846c8974d

millerdev commented:
@esoergel wrote:
> There's a build failure I haven't looked at

Yeah, sorry the PR description wasn't very informative. I'm using this PR to find and fix test failures caused by the upgrade.

millerdev committed:
Work around buggy databases = '__all__'

millerdev committed:
Refactor: simplify decorator

millerdev committed:
Do not transact shard dbs in tests

millerdev committed:
Work around Django 2.1 db router changes

Changes from https://github.com/django/django/pull/9360 caused more aggressive router lookups, which resulted in infinite recursion when attempting to save some new related objects.

millerdev committed:
Fix ValueError: seek of closed file

Caused by https://github.com/django/django/pull/8310

millerdev committed:
Fix partitioned TransactionTestCase

millerdev committed:
Improve diff output in travis

Troubleshooting test failure: custom.icds_reports.tests.agg_tests.reports.test_awc_reports:TestAWCReport.test_awc_reports_pse_charts_1
So far have been unable to run the test locally.

millerdev committed:
Decimal to float in tests

millerdev committed:
Merge remote-tracking branch 'origin/master' into dm/dj2-fix-tests

millerdev committed:
DO NOT MERGE skip failing test

millerdev committed:
Merge branch 'dm/citus-test-db' into dm/dj2-fix-tests

millerdev committed:
Detect proxy dbs from new-style proxy config

millerdev committed:
Refactor: simplify get attribute

dannyroberts commented:
Daniel and I offlined and I'm getting us set up to QA this even before all tests pass.

@millerdev there was an issue deploying to staging, which I was also able to repro locally on this branch:

```
$ ./manage.py compilejsi18n
Traceback (most recent call last):
  File "./manage.py", line 189, in <module>
    main()
  File "./manage.py", line 41, in main
    execute_from_command_line(sys.argv)
  File "/Users/danielroberts/.virtualenvs/cchq/lib/python3.7/site-packages/django/core/management/__init__.py", line 381, in execute_from_command_line
    utility.execute()
  File "/Users/danielroberts/.virtualenvs/cchq/lib/python3.7/site-packages/django/core/management/__init__.py", line 375, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/Users/danielroberts/.virtualenvs/cchq/lib/python3.7/site-packages/django/core/management/__init__.py", line 224, in fetch_command
    klass = load_command_class(app_name, subcommand)
  File "/Users/danielroberts/.virtualenvs/cchq/lib/python3.7/site-packages/django/core/management/__init__.py", line 36, in load_command_class
    module = import_module('%s.management.commands.%s' % (app_name, name))
  File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.7/lib/python3.7/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 677, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 728, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/Users/danielroberts/.virtualenvs/cchq/lib/python3.7/site-packages/statici18n/management/commands/compilejsi18n.py", line 11, in <module>
    from django.views.i18n import (get_javascript_catalog,
ImportError: cannot import name 'get_javascript_catalog' from 'django.views.i18n' (/Users/danielroberts/.virtualenvs/cchq/lib/python3.7/site-packages/django/views/i18n.py)
```

millerdev committed:
Fix test transaction management for icds_reports

Aggregation tests use DDL to drop and create tables, which does not work with django.test.TestCase transaction management because multiple DDL statements cannot be executed in the context of a single transaction.

millerdev committed:
Update django-two-factor-auth and django-otp

millerdev committed:
Fix db router

corehq.apps.commtrack.tests.test_xml.CommTrackArchiveSubmissionTestSQL.test_archive_last_form was failing with

django.db.utils.IntegrityError: insert or update on table "form_processor_ledgervalue" violates foreign key constraint "cd40c15ceaad5d793e09d0b69eb4ed88"
DETAIL:  Key (case_id)=(...) is not present in table "form_processor_commcarecasesql".

millerdev committed:
Handle django-two-factor-auth urlpatterns tuples

Fixes error on ./manage.py runserver:

SystemCheckError: System check identified some issues:

ERRORS:
?: (urls.E004) Your URL pattern 'two_factor' is invalid. Ensure that urlpatterns is a list of path() and/or re_path() instances.
	HINT: Try removing the string 'two_factor'. The list of urlpatterns should not have a prefix string as the first element.
?: (urls.E004) Your URL pattern 'two_factor_twilio' is invalid. Ensure that urlpatterns is a list of path() and/or re_path() instances.
	HINT: Try removing the string 'two_factor_twilio'. The list of urlpatterns should not have a prefix string as the first element.
?: (urls.E004) Your URL pattern [<URLPattern '^account/login/$' [name='login']>, <URLPattern '^account/two_factor/setup/$' [name='setup']>, <URLPattern '^account/two_factor/qrcode/$' [name='qr']>, <URLPattern '^account/two_factor/setup/complete/$' [name='setup_complete']>, <URLPattern '^account/two_factor/backup/tokens/$' [name='backup_tokens']>, <URLPattern '^account/two_factor/backup/phone/register/$' [name='phone_create']>, <URLPattern '^account/two_factor/backup/phone/unregister/(?P<pk>\d+)/$' [name='phone_delete']>, <URLPattern '^account/two_factor/$' [name='profile']>, <URLPattern '^account/two_factor/disable/$' [name='disable']>] is invalid. Ensure that urlpatterns is a list of path() and/or re_path() instances.
?: (urls.E004) Your URL pattern [<URLPattern '^twilio/inbound/two_factor/(?P<token>\d+)/$' [name='call_app']>] is invalid. Ensure that urlpatterns is a list of path() and/or re_path() instances.

millerdev committed:
Backward compatible django-two-factor-auth urlpatterns

millerdev committed:
Fix XFormManagementView.get_xform_ids

AttributeError: 'OTPMiddleware' object has no attribute 'process_request'

millerdev committed:
Another attempt at test case databases

Resolves AssertionError: Database queries to 'proxy' are not allowed in this test. Add 'proxy' to corehq.form_processor.tests.test_reprocess_errors.TestTransactionErrors.databases to ensure proper test isolation and silence this failure.

millerdev committed:
Dynamic transaction_exempt_databases

Resolves error in TestCase._enter_atomics:
AssertionError: Database queries to 'icds-ucr' are not allowed in this test. Add 'icds-ucr' to corehq.apps.accounting.tests.test_autopay.TestBillingAutoPay.databases to ensure proper test isolation and silence this failure.

millerdev committed:
Dynamic TestCase.databases

Resolves errors in ICDS tests:
AssertionError: Database queries to 'icds-ucr' are not allowed in this test. Add 'icds-ucr' to custom.icds.tests.test_aww_vhnd_survey.TestAWWVHNDSurveyIndicator.databases to ensure proper test isolation and silence this failure.

millerdev commented:
This PR cannot be merged because of https://github.com/dimagi/commcare-hq/pull/27911/commits/f3b26677a88a01dc0240a752e063cfbde87fc072. All other commits have been cherry-picked into separate PRs, some of which are backward compatible and have been merged, and one that contains non-backward-compatible changes.

The branch has not been deleted yet because the ICDS team is still basing some work on it.
